rust/crypto/src/hash.rs

```diff
@@ -0,0 +1,80 @@
+//
+// Copyright 2021 Signal Messenger, LLC.
+// SPDX-License-Identifier: AGPL-3.0-only
+//
+
+use crate::{Error, Result};
+
+use hmac::{Hmac, Mac, NewMac};
+use sha1::Sha1;
+use sha2::{Digest, Sha256, Sha512};
+
+#[derive(Clone)]
+pub enum CryptographicMac {
+    HmacSha256(Hmac<Sha256>),
+    HmacSha1(Hmac<Sha1>),
+}
+
+impl CryptographicMac {
+    pub fn new(algo: &str, key: &[u8]) -> Result<Self> {
+        match algo {
+            "HMACSha1" | "HmacSha1" => Ok(Self::HmacSha1(
+                Hmac::<Sha1>::new_varkey(key).expect("HMAC accepts any key length"),
+            )),
+            "HMACSha256" | "HmacSha256" => Ok(Self::HmacSha256(
```

Comment:
I was going to suggest that the string-based generic entry points be implemented in Java, since they're not a design we'd use for any other languages even if we want to expose the operations, but it does come out really compact in Rust with the enum. The alternate approach would probably be to use a trait instead, and have the bridge functions return `Box<dyn CryptographicMac>`. Thoughts?

Reply:
I'd rather avoid the extra heap allocation if we can, given how small most of these objects are.